File: ts/test/fixture/transport/context.py
from demo_sdk.core.context import DemoContext


root = DemoContext({"config": {}})
entity = DemoContext({}, root)
first = DemoContext({"opname": "list"}, entity)
first.ctrl.paging = {"cursor": "first"}
second = DemoContext({"opname": "list"}, entity)
assert first.ctrl is not second.ctrl
assert root.ctrl.paging is None
assert second.ctrl.paging is None
assert DemoContext({}, first).ctrl is first.ctrl
paging = {"cursor": "explicit"}
explicit = DemoContext({"opname": "list", "ctrl": {"paging": paging}}, entity)
assert explicit.ctrl.paging is paging
print("context: isolated defaults; explicit and nested controls preserved")

File: ts/test/fixture/transport/context.php
<?php
require_once __DIR__ . '/utility/struct/Struct.php';
require_once __DIR__ . '/core/Context.php';

$root = new DemoContext(['config' => []]);
$entity = new DemoContext([], $root);
$first = new DemoContext(['opname' => 'list'], $entity);
$first->ctrl->paging = ['cursor' => 'first'];
$second = new DemoContext(['opname' => 'list'], $entity);
if ($first->ctrl === $second->ctrl) throw new Exception('shared operation control');
if ($root->ctrl->paging !== null) throw new Exception('root paging changed');
if ($second->ctrl->paging !== null) throw new Exception('paging leaked');
if ((new DemoContext([], $first))->ctrl !== $first->ctrl) throw new Exception('nested control lost');
$explicit = new DemoContext(['opname' => 'list', 'ctrl' => ['paging' => ['cursor' => 'explicit']]], $entity);
if ($explicit->ctrl->paging['cursor'] !== 'explicit') throw new Exception('explicit paging lost');
echo "context: isolated defaults; explicit and nested controls preserved\n";
